Add type annotation logic for complex numbers

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51145 Support complex number list in JIT
* #50885 Add serialization logic for complex numbers
* **#50884 Add type annotation logic for complex numbers**

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D26086963](https://our.internmc.facebook.com/intern/diff/D26086963)

anjali411 committed:
Add type annotation logic for complex numbers

[ghstack-poisoned]

anjali411 committed:
Update on "Add type annotation logic for complex numbers"




[ghstack-poisoned]

anjali411 committed:
Update on "Add type annotation logic for complex numbers"




[ghstack-poisoned]

anjali411 committed:
Update on "Add type annotation logic for complex numbers"




[ghstack-poisoned]

anjali411 committed:
Update on "Add type annotation logic for complex numbers"


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

anjali411 committed:
Update on "Add type annotation logic for complex numbers"


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

anjali411 committed:
Update on "Add type annotation logic for complex numbers"


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

SplitInfinity commented on test/test_complex.py:
```diff
@@ -1,9 +1,22 @@
 
 devices = (torch.device('cpu'), torch.device('cuda:0'))
 
+class TestComplexJIT(JitTestCase):
```
```suggestion
class TestComplex(JitTestCase):
```

SplitInfinity commented on test/test_complex.py:
```diff
@@ -1,9 +1,22 @@
+
+        b = 3 + 5j
+        result = self.checkScript(fn, (b,))
+        cu = torch.jit.CompilationUnit('''
+                def fn(a: complex):
+                    return a
+        ''')
```
`checkScript` should test this internally as well, so I think you can delete this.

SplitInfinity commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -229,6 +233,9 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
         return py::cast<int64_t>(obj);
       } else if (py::isinstance<py::float_>(obj)) {
         return py::cast<double>(obj);
+      } else if (PyComplex_CheckExact(obj.ptr())) {
```
Dumb question but why does this line not say something like `py::isinstance<py::complex_>(...)` like all the others?

anjali411 replied:
Complex type is not available in pybind (see doc here: https://pybind11.readthedocs.io/en/stable/advanced/pycpp/object.html#accessing-python-libraries-from-c) so I just pass the underlying `PyObject*` to `PyComplex_CheckExact` to check whether it's complex

SplitInfinity commented on torch/jit/annotations.py:
```diff
@@ -305,6 +305,8 @@ def try_ann_to_type(ann, loc):
         return FutureType(try_ann_to_type(ann.__args__[0], loc))
     if ann is float:
         return FloatType.get()
+    if inspect.isclass(ann) and issubclass(ann, builtins.complex):
+        return ComplexDoubleType.get()
```
Another dumb question, why is this not `ann is complex`?

SplitInfinity replied:
Also, is there a usecase for creating subclasses of `complex`?

anjali411 replied:
complex is a python class hence the `isclass` call. umm no there's no use specific case of creating subclasses of complex

SplitInfinity replied:
But aren't `int` and `float` classes too?

anjali411 replied:
ok wait you are right `ann is complex` works!

SplitInfinity commented on torch/csrc/jit/frontend/function_schema_parser.cpp:
```diff
@@ -248,6 +248,7 @@ struct SchemaParser {
       case TypeKind::IntType:
       case TypeKind::BoolType:
       case TypeKind::FloatType:
+      case TypeKind::ComplexDoubleType:
         return parseSingleConstant(arg_type->kind());
```
`parseSingleConstant` needs to be updated to handle complex numbers. I would either get rid of it and do it in a follow-up PR or add a test for it (and update `parseSingleConstant` and other code as needed).

anjali411 replied:
I'll just remove this change for the current PR and add it separately

SplitInfinity reviewed (commented):
Nice! Do you plan to support constant literals in TorchScript code in this diff? If not, I think it is ready to land. If so, I think some more changes are required.

anjali411 committed:
Update on "Add type annotation logic for complex numbers"


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

anjali411 commented:
> Nice! Do you plan to support constant literals in TorchScript code in this diff? If not, I think it is ready to land. If so, I think some more changes are required.

no I'll just make a separate PR for that! also, addressed all the comments and updated the PR

SplitInfinity reviewed (approved):
🚀
